### DISCUSSION TITLE
multi_body_system_to_urdf help

### DISCUSSION THREAD

**Discussion by The-Odor** [ci-group/revolve2#433] [Ask for help]
I am trying to use [multi_body_system_to_urdf](https://github.com/ci-group/revolve2/blob/b345df58e4bdaa532bb78558b5286512b2c016e2/simulation/revolve2/simulation/scene/conversion/_multi_body_system_to_urdf.py#L16), but I can't find that it is made too clear how to use it. I've followed the tracks of imports up to one of your example files through your github and have tried to replicate how it is called. I ended up making a scene, turning it into a batch, simulating the batch, extracting the scene from the batch, and as far as I can tell from the github examples, there should now be an attribute called scene.multi_body_system, which I should feed into multi_body_system_to_urdf, but I can't find that attribute. I've tried excluding some of the steps, but that didn't work, so I'm trying to make it work by looking at it in a form of call that is as close as I can make it by looking at its main call and following the breadcrumbs up

**Comment by surgura** [comment]
Unless you are trying to do soomething very specific, this is an internal function that you should not use. What are you trying to achieve? On Jan 22, 2024 15:57, The-Odor ***@***.***> wrote:
I am trying to use multi_body_system_to_urdf, but I can't find that it is made too clear how to use it. I've followed the tracks of imports up to one of your example files through your github and have tried to replicate how it is called. I ended up making a scene, turning it into a batch, simulating the batch, extracting the scene from the batch, and as far as I can tell from the github examples, there should now be an attribute called scene.multi_body_system, which I should feed into multi_body_system_to_urdf, but I can't find that attribute. I've tried excluding some of the steps, but that didn't work, so I'm trying to make it work by looking at it in a form of call that is as close as I can make it by looking at its main call and following the breadcrumbs up

**Reply by The-Odor** [reply]
I am trying to translate your robot structures into URDF for use in Unity

**Reply by surgura** [reply]
I see. We have not taken that use case into account. The URDF might be very incomplete; some features like motors and sensors are not transferred well. For our system it is more of an intermediate step that we convert to MJCF(mujoco format) and then manually complete it. If you really want to obtain that incomplete URDF, take a look at 'BodyToMultiBodySystemConverter' and then 'multi_body_system_to_urdf'. Those two functions are enough to obtain the URDF string from a modular robot body.

**Reply by The-Odor** [reply]
Fantasic! I think that may have been precisely what I needed!

```
from revolve2.modular_robot_simulation._build_multi_body_systems import BodyToMultiBodySystemConverter
from revolve2.simulation.scene import Pose
from revolve2.ci_group import modular_robots_v1
from revolve2.simulation.scene.conversion._multi_body_system_to_urdf import multi_body_system_to_urdf

def main():
    converter = BodyToMultiBodySystemConverter()

    body = modular_robots_v1.gecko_v1()
    pose = Pose()
    translate_z_aabb = True
    MBS, _ = converter.convert_robot_body(body=body, pose=pose, translate_z_aabb=translate_z_aabb)

    print(multi_body_system_to_urdf(MBS, "test"))

```
Prints out something that looks like URDF!
